### PR TITLE
Alerting: Show affected alert rules when creating Silence

### LINKED ISSUE #35725: Alerting: preview alerts in silence form
**What would you like to be added**:

In the silence form, it should be possible to preview matching alerts before saving.

**Why is this needed**:
It would be helpful to see what alerts will be affected by the silence

### PR DESCRIPTION
**Pull request by peterholmberg**
**What this PR does / why we need it**:
Adds a table that shows which alert rules will be silenced by a silence by matching labels.

Also in this PR:
* Aligned the different sections so their width matches. 
* Removing createdBy field
* Increase height of description field
* ~TODO: Add default description. Something like: Silencing rules with matchers etc.~

**Which issue(s) this PR fixes**:
Fixes https://github.com/grafana/grafana/issues/35725

**Special notes for your reviewer**:

### CODE REVIEW

File: public/app/features/alerting/unified/utils/matchers.ts [hunk 1]
```diff
@@ -24,3 +26,36 @@ export const getMatcherQueryParams = (labels: Labels) => {
 
   return matcherUrlParams;
 };
+
+export const findAlertRulesWithMatchers = (rules: CombinedRule[], matchers: MatcherFieldValue[]) => {
```
Comment (gillesdemey) [inline public/app/features/alerting/unified/utils/matchers.ts:30]: I think this one could use a proper return type interface

**Comment by gillesdemey** [comment]
I think we should add an empty state when no alerts are matching the labels to make it more explicit that there are no matches, wdyt?

File: public/app/features/alerting/unified/components/silences/MatchedSilencedRules.tsx [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+      setMatchedAlertRules(matchedRules);
+    },
+    500,
+    [combinedNamespaces, matchers]
+  );
+
+  return (
+    <div>
+      <h4>
+        Affected alerts {matchedAlertRules && matchedAlertRules.length > 0 ? `(${matchedAlertRules.length})` : ''}
```
Comment (gillesdemey) [inline public/app/features/alerting/unified/components/silences/MatchedSilencedRules.tsx:52]: I'm not too familiar with the design system; but perhaps a `<Badge>` component could be used here to set the value apart from the header?

image

File: public/app/features/alerting/unified/components/silences/MatchedSilencedRules.tsx [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+import { CombinedRule } from 'app/types/unified-alerting';
+import { MatcherFieldValue, SilenceFormFields } from '../../types/silence-form';
+
+type MatchedRulesTableItemProps = DynamicTableItemProps<{
+  matchedRule: CombinedRule;
+}>;
+type MatchedRulesTableColumnProps = DynamicTableColumnProps<{ matchedRule: CombinedRule }>;
+
+export const MatchedSilencedRules = () => {
+  const [matchedAlertRules, setMatchedAlertRules] = useState<MatchedRulesTableItemProps[]>();
```
Comment (gillesdemey) [inline public/app/features/alerting/unified/components/silences/MatchedSilencedRules.tsx:23]: If we set the default value to `[]` we can omit most of the `matchedAlertRules &&` and `matchedAlertRules?` checks

**Review by gillesdemey — COMMENTED** [review]
Just a few nits

File: public/app/features/alerting/unified/utils/matchers.ts [hunk 4]
```diff
@@ -24,3 +26,36 @@ export const getMatcherQueryParams = (labels: Labels) => {
...
+  const hasMatcher = (rule: CombinedRule, matcher: MatcherFieldValue) => {
+    return Object.entries(rule.labels).some(([key, value]) => {
```
Comment (konrad147) [inline public/app/features/alerting/unified/utils/matchers.ts:32]: There is a very similar piece of code in the `labelsMatchMatchers` function from `public/app/features/alerting/unified/utils/alertmanager.ts`. Maybe it's worth reusing it?

File: public/app/features/alerting/unified/components/silences/MatchedSilencedRules.tsx [hunk 5]
```diff
@@ -48,7 +48,12 @@ export const MatchedSilencedRules = () => {
 
   return (
     <div>
-      <h4>Affected alerts {matchedAlertRules.length > 0 ? `(${matchedAlertRules.length})` : ''}</h4>
+      <h4 className={styles.title}>
+        Affected alerts
+        {matchedAlertRules.length > 0 ? (
+          <Badge className={styles.badge} color="blue" text={matchedAlertRules.length} />
+        ) : null}{' '}
```
Comment (gillesdemey) [inline public/app/features/alerting/unified/components/silences/MatchedSilencedRules.tsx:55]: Prettier seems to love to add these `{' '}` 👀

Comment (peterholmberg) [reply public/app/features/alerting/unified/components/silences/MatchedSilencedRules.tsx:55]: 😩

File: public/app/features/alerting/unified/utils/matchers.ts (hunk 4 above)
Comment (peterholmberg) [reply public/app/features/alerting/unified/utils/matchers.ts:39]: Nice! Although it did not fit this use case.

**Review by gillesdemey — APPROVED** [review]

**Review by konrad147 — APPROVED** [review]

**Pull request merged**